src/Agent/NewRelic/Agent/Core/Segments/Segment.cs

```diff
@@ -441,5 +441,26 @@ public string GetCategory()
         {
             return EnumNameCache<SpanCategory>.GetName(Data.SpanCategory);
         }
+
+        public static int? CheckAllSameThread(ICollection<Segment> segments)
```

Comment:
Would it be safer to check if a segment was created from an instrumented method where the IsAsync flag is set? I think that the exclusive time calculation is mostly complicated when async-await is used, and when async-await is used the IsAsync flag should be set.

Reply:
That was my first approach, but we don't keep track of `IsAsync` beyond the initial check, and I had to touch a lot of code to get it to propagate down to where the attributes are set. This way I thought was at least similar to how we track parent/child relationships. Maybe we can brainstorm some alternatives.

Reply:
This will act as a heuristic because it is just checking that each segment started on the same thread. It will not successfully ignore all of the async-await scenarios, because the thread will not change until there is an await. This is one of the reasons why the exclusive time calculation is comparing the the starting thread of the parent to the ending thread id of the child.

Reply:
Ok, I think I'm with you now. I misread the conversation in the spec PR and also misunderstood how we were tracking thread IDs for Segments. Is this more in line with what you were expecting?

Reply:
Yes, I think that your changes will better detect the case where code is async even if the async methods started on the same thread.